### ISSUE TITLE
PR 2 — `Outputs` dataclass + `PreprocInfo` struct

### ISSUE THREAD

**Issue by gitttt-1234** [talmolab/sleap-nn#510]
**Parent epic:** #508
**Prerequisites:** #509
**Blocks:** #512, #514, #516

---

## Goal

Replace dict-based inference outputs with the structured `Outputs` dataclass, and introduce `PreprocInfo` — a struct that carries inverse-transform metadata from `preprocess()` to `postprocess()`.

Currently every `InferenceModel.forward()` returns a `Dict[str, torch.Tensor]` with keys like `pred_instance_peaks`, `pred_peak_values`, `pred_confmaps`, `instance_bbox`, etc. The keys are typo-prone, the shapes are undocumented, and there's no place to hang methods like `.plot()` or `.slim()`. This PR replaces the dict with a typed container.

`PreprocInfo` is new: it carries `eff_scale`, `input_scale`, `output_stride`, `pad_amount`, `crop_offsets` between preprocess and postprocess so the coordinate ladder in `ops/coord.py` (#509) can be applied in one place.

### `PreprocInfo`

```python
# sleap_nn/inference/layers/base.py  (config portion)

from typing import Optional, Tuple
import attrs
import torch


@attrs.define
class PreprocInfo:
    """Metadata from preprocessing, used to reverse coordinate transforms.

    Produced by InferenceLayer.preprocess(); consumed by .postprocess().
    """

    original_size: Tuple[int, int] = (0, 0)
    processed_size: Tuple[int, int] = (0, 0)
    eff_scale: float = 1.0
    input_scale: float = 1.0
    output_stride: int = 1
    pad_amount: Tuple[int, int] = (0, 0)
    crop_offsets: Optional[torch.Tensor] = None
```

### `Outputs`

Full canonical definition from design doc §4B.4 — reproduced here for reviewer convenience:

```python
# sleap_nn/inference/outputs.py

from typing import Any, Dict, List, Optional, Tuple, Union
import attrs
import numpy as np
import torch
import sleap_io as sio

from sleap_nn.inference.layers.base import PreprocInfo


@attrs.define
class Outputs:
    """Structured container for inference outputs.

    Shape convention:
        B = batch size, I = max instances, N = nodes, C = classes,
        H, W = spatial dims, E = number of edges.
        NaN indicates missing/invalid predictions.
    """

    # ── Images (optional; None unless explicitly requested) ──────────
    original_image: Optional[torch.Tensor] = None        # (B, C, H, W), uint8/float32
    processed_image: Optional[torch.Tensor] = None       # (B, C, H', W'), float32
    crops: Optional[torch.Tensor] = None                 # (B, I, C, cH, cW); top-down only

    # ── Core predictions ─────────────────────────────────────────────
    pred_keypoints: Optional[torch.Tensor] = None        # (B, I, N, 2) original-image (x, y)
    pred_peak_values: Optional[torch.Tensor] = None      # (B, I, N)
    pred_confmaps: Optional[torch.Tensor] = None         # (B, N, H, W) — heavy, opt-in
    pred_pafs: Optional[torch.Tensor] = None             # (B, 2E, H, W) — heavy, opt-in
    pred_centroids: Optional[torch.Tensor] = None        # (B, I, 2)
    pred_centroid_values: Optional[torch.Tensor] = None  # (B, I)

    # ── Instance-level metadata ──────────────────────────────────────
    instance_scores: Optional[torch.Tensor] = None       # (B, I)
    instance_valid: Optional[torch.Tensor] = None        # (B, I), bool
    instance_bboxes: Optional[torch.Tensor] = None       # (B, I, 4, 2)

    # ── Multi-class predictions ──────────────────────────────────────
    pred_class_vectors: Optional[torch.Tensor] = None    # (B, I, N, C)
    pred_class_maps: Optional[torch.Tensor] = None       # (B, C, H, W)
    pred_class_inds: Optional[torch.Tensor] = None       # (B, I, N)
    pred_class_probs: Optional[torch.Tensor] = None      # (B, I, C)

    # ── PAF graph (opt-in) ───────────────────────────────────────────
    pred_paf_graph: Optional[Tuple[torch.Tensor, ...]] = None

    # ── Preprocessing metadata ───────────────────────────────────────
    preprocess_info: Optional[PreprocInfo] = None

    # ── Frame/video metadata ─────────────────────────────────────────
    frame_indices: Optional[torch.Tensor] = None         # (B,), int64
    video_indices: Optional[torch.Tensor] = None         # (B,), int64

    # Methods: to, cpu, detach, numpy, as_dict, slim, drop_heavy,
    #         to_instances, to_labels, plot
    # Properties: batch_size, n_instances, n_nodes
    # NOTE: NO filter methods. All filtering lives in FilterConfig +
    #       FilterPipeline (#516). Outputs is a pure data container.
    # See design doc §4B.4 for full signatures.
```

## Scope decisions

**Full method implementations in this PR:**
- `to(device)`, `cpu()`, `detach()`, `numpy()`, `as_dict()` — straightforward attrs.evolve patterns
- `slim()`, `drop_heavy()` — needed by `IncrementalLabelsWriter` in #516
- Properties `batch_size`, `n_instances`, `n_nodes`

**Stubs (`NotImplementedError` with TODO) in this PR:**
- `to_instances()` — depends on final coord space decisions from #514
- `to_labels()` — depends on `Predictor` shape from #516
- `plot()` — can land as stub; full matplotlib impl is low-risk follow-up

**Explicitly NOT in this PR (or ever, on Outputs):**
- No `filter_by_confidence()`, no `filter()`, no filtering methods of any kind on Outputs. All filtering is concentrated in `FilterConfig` + `FilterPipeline` — see #516 for those. This is a design decision (design doc §4B.4): Outputs stays a pure data container.

## Pickle-safety contract

`Outputs.slim()` is the **canonical cross-process boundary**. Anything that crosses an `mp.Queue` (e.g., the PAF worker pool in #517) must call `.slim()` first. `slim()` guarantees:

1. All tensors are on CPU (`.cpu()`)
2. All tensors are detached from autograd (`.detach()`)
3. Heavy tensors (`pred_confmaps`, `pred_pafs`, `original_image`, `processed_image`, `crops`) are dropped
4. `pred_paf_graph` (tuple of tensors) is either dropped or each tensor is `.cpu().detach()`
5. `PreprocInfo.crop_offsets` (if a tensor) is `.cpu().detach()`

Result: `pickle.dumps(outputs.slim())` and round-trip `pickle.loads(pickle.dumps(...))` succeed without any `RuntimeError` (no CUDA storage handoff, no `grad_fn` references). This is enforced by a unit test, not just convention.

## Acceptance criteria

- [ ] `sleap_nn/inference/outputs.py` with full field list and documented shapes
- [ ] `PreprocInfo` added to `sleap_nn/inference/layers/base.py` (new file — creates the `layers/` package)
- [ ] All listed "full impl" methods have tests (round-trip to/from numpy, drop_heavy drops the right fields, properties return correct values on fixture Outputs)
- [ ] Stubs raise `NotImplementedError("TODO: land in PR <N>")` with the correct PR number
- [ ] **Pickle round-trip test:** `pickle.loads(pickle.dumps(outputs.slim()))` returns an `Outputs` equal to the original, on CPU, with no `grad_fn` and no heavy tensors. Run on a fixture `Outputs` constructed on each available device (CPU + MPS, plus CUDA when available).
- [ ] **Pickle without `.slim()` is allowed to fail loud:** test that `pickle.dumps(outputs_with_grad_fn)` either succeeds or raises a clear error — never silently produces a corrupt payload.
- [ ] `ruff check` + `black --check` clean

## Files touched

**Created:**
- `sleap_nn/inference/outputs.py`
- `sleap_nn/inference/layers/__init__.py`
- `sleap_nn/inference/layers/base.py` (only `PreprocInfo` for now; full `InferenceLayer` lands in #512)
- `tests/inference/test_outputs.py`

## Out of scope

- No integration with any existing code — this PR only introduces the types. Integration starts at #512.
- No `.plot()` matplotlib implementation (stub only; full impl is a follow-up)

**Comment by gitttt-1234** [comment]
Closed — merged as part of the inference refactor stack (epic #508).
